### PR TITLE
cppzmq: new wrap

### PR DESCRIPTION
**Pull request by stephanlachnit**
This currently still requires libzmq, which I will add as wrap hopefully soon.

### CODE REVIEW

**Comment by stephanlachnit** [comment]
@eli-schwartz @jpakkane can we merge this already? I've made some progress on libzmq (https://github.com/stephanlachnit/libzmq/tree/p-meson), but it selecting the correct option will take some more time. Doesn't block cppzmq though as it works fine with system libzmq.

**Comment by eli-schwartz** [comment]
Sorry, I forgot to look at it because I assumed that you were going to be updating the PR to also wrap zeromq.

I think you should install zeromq for macOS via `brew_packages`. Additionally, how about building and running the tests/ directory?

**Comment by stephanlachnit** [comment]
> I think you should install zeromq for macOS via brew_packages.

Good point

> Additionally, how about building and running the tests/ directory?

Build the tests, could also add the examples. Looks like the CI doesn't take catch2 from wrapdb, how do I do this?

Is there some kind of preferred way to enable testing in meson? I think when using cppzmq in a subproject, the tests and examples should not be build by default.

**Comment by eli-schwartz** [comment]
Just add a meson_options.txt with a tests feature option, and use that as the required kwarg for catch2. If it is required, then it will be picked up via wraps, wrap fallbacks don't automatically get used for optional dependencies.

(There's a kwarg to allow fallback anyway.)

**Comment by stephanlachnit** [comment]
Ah right, I see. But then the feature would have to be enabled by default for this to work, right?

A bit offtopic, but maybe meson should add a builtin tests bool? That could be disabled for subprojects by default, but e.g. here in the CI enabled via cli.

**Comment by neheb** [comment]
That wouldn't make much sense. Note that you still have to run ninja tests or meson tests to get them to run. A simple compile does that but does not run or install them.

File: subprojects/packagefiles/cppzmq/meson.build [hunk 1]
```diff
@@ -1,10 +1,39 @@
...
 )
 
 libzmq_dep = dependency('libzmq')
 
 cppzmq_inc = include_directories('.')
 
 cppzmq_dep = declare_dependency(include_directories: cppzmq_inc, dependencies: libzmq_dep)
+
+# tests
+catch2_dep = dependency('catch2', required: false)
```
Comment (neheb) [inline subprojects/packagefiles/cppzmq/meson.build:13]: Should probably use a feature option here to allow disabling the tests.

**Comment by stephanlachnit** [comment]
> That wouldn't make much sense.

I disagree, tests should only be compiled if wished by the user (developer vs user). Sure, enabling them by default is a good default, but having a default way to not compile tests is something useful. It might not make a difference here, but if you have a giant testsuite containing maybe hundreds of files, it can take quite a while to compile.
Lots of CMake projects have it, and even some meson projects (e.g. https://github.com/libratbag/piper/blob/7f9834cc5901f790a596bae61e3eb81e63d7c2bd/meson_options.txt#L1-L4).

> Note that you still have to run ninja tests or meson tests to get them to run. A simple compile does that but does not run or install them.

Sure, but I have my own tests defined, I don't care about tests from the subproject - possible test failures have nothing to do with my project.

**Comment by eli-schwartz** [comment]
> I disagree, tests should only be compiled if wished by the user (developer vs user). Sure, enabling them by default is a good default, but having a default way to not compile tests is something useful.

Actually, this shouldn't be an option, and enabling to compile them by default is a pointless waste.

It's just a bug in Meson that currently, `build_by_default: false` is ignored for testsuite executables.

What should happen is you mark every test executable to not build by default, but it's fine to unconditionally define their rules. Running `ninja test` should build the testsuite executables, but `ninja all` shouldn't build them.

This works fine in most cases, though if the tests require additional dependencies then it's probably desirable to have an option for that.

**Comment by stephanlachnit** [comment]
> What should happen is you mark every test executable to not build by default, but it's fine to unconditionally define their rules. Running ninja test should build the testsuite executables, but ninja all shouldn't build them.

Ah thanks! This seems like a good idea - but there is still the subproject problem. But this should be fixable with an option for `meson test`, i.e. `meson test --with-subprojects`.

**Comment by eli-schwartz** [comment]
```
meson test --suite mainprojectname
```

**Comment by stephanlachnit** [comment]
> meson test --suite mainprojectname

Thanks!

**Comment by stephanlachnit** [comment]
Added tests and examples with `build_by_default: false`, so from my side good to merge now.

> Just add a meson_options.txt with a tests feature option, and use that as the required kwarg for catch2. If it is required, then it will be picked up via wraps, wrap fallbacks don't automatically get used for optional dependencies.

Did that, it still is not picked up by the CI when left on auto.

File: subprojects/packagefiles/cppzmq/meson.build [hunk 2]
```diff
@@ -0,0 +1,62 @@
+project('cppzmq', 'cpp',
+  version : '4.8.1',
+  license : 'MIT',
+)
+
+libzmq_dep = dependency('libzmq')
+
+cppzmq_inc = include_directories('.')
+
+cppzmq_dep = declare_dependency(include_directories: cppzmq_inc, dependencies: libzmq_dep)
```
Comment (eli-schwartz) [inline subprojects/packagefiles/cppzmq/meson.build:10]: Can we just inline the inc here?

File: subprojects/packagefiles/cppzmq/meson.build [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+# examples
+thread_dep = dependency('threads')
+
+ex_pubsub_multithread_inproc_exe = executable('pubsub_multithread_inproc',
+  'examples/pubsub_multithread_inproc.cpp',
```
Comment (eli-schwartz) [inline subprojects/packagefiles/cppzmq/meson.build:16]: I was wondering why I could not build this until I saw https://github.com/zeromq/cppzmq/pull/568

An absolutely filthy hack: 

```suggestion
  'examples/pubsub_multithread_inproc.cpp',
  # fixed upstream in https://github.com/zeromq/cppzmq/pull/568
  cpp_args: ['-include', 'thread'],
```

File: subprojects/packagefiles/cppzmq/meson.build [hunk 4]
```diff
@@ -0,0 +1,62 @@
+project('cppzmq', 'cpp',
+  version : '4.8.1',
+  license : 'MIT',
```
Comment (eli-schwartz) [inline subprojects/packagefiles/cppzmq/meson.build:3]:
```meson
meson_version: '>=0.47.0'
```

**Review by eli-schwartz — COMMENTED** [review]
The ci_config.json can pass `-Dcppzmq:catch2=required` as build_options for this wrap. This will guarantee fallback happens.

It's also possible to use `dependency(..., allow_fallback: true)` but that bumps the minimum version of Meson from 0.47.0 to 0.56.0

File: subprojects/packagefiles/cppzmq/meson.build (hunk 3 above)
Comment (stephanlachnit) [reply subprojects/packagefiles/cppzmq/meson.build:14]: Ah thanks, that's easier than adding patch files.

**Comment by stephanlachnit** [comment]
> The ci_config.json can pass `-Dcppzmq:catch2=required` as build_options for this wrap. This will guarantee fallback happens.

`cppzmq:catch2=enabled` but thanks, done.

**Comment by stephanlachnit** [comment]
@eli-schwartz any more objections or can this be merged?

**Pull request merged**